Events.md

```diff
@@ -0,0 +1,225 @@
+
+We tried to list the main points that we would like to do with events.
+
+1. Use as little gas as possible
+2. An indexer should be able to keep track of the state of Mangrove.
+3. Doing RPC calls directly, should be able to find offers and other information based on offerlist, maker, taker, offer id, etc.
+
+These 3 points all have there own direction and it is therefore not possible to find a solution that is optimal for all 3 points.
+We have therefore tried to find a solution that is a good balance between the all 3 points.
+
+## Events
+
+### Approval
+
+This is emitted when a user permits another address to use a certain amount of its funds to do market orders, or when a user revokes another address to use a certain amount of its funds.
+
+Approvals are based on the pair of outbound and inbound token. Be aware that it is not offerlist bases, as an offerlist also holds the tickscale.
+
+We emit `outbound` token, `inbound` token, `owner`, msg.sender (`spender`), `value`. Where `owner` is the one who owns the funds, `spender` is the one who is allowed to use the funds and `value` is the amount of funds that is allowed to be used.
+
+Outbound, inbound and owner is indexed, this way one can filter on these fields when doing RPC calls. If we could somehow combine outbound and inbound into one field, then we could also index the spender.
+
+### Credit
+
+This is emitted when a user's account on Mangrove is credited with some native funds, to be used as provision for offers.
+
+It emits the `maker`'s address and the `value` credited
+
+The `maker` address is indexed so that we can filter on it when doing RPC calls.
+
+These are the scenarios where it can happen:
+
+- Fund Mangrove directly
+- Fund Mangrove when posting an offer
+- When updating an offer
+  - Funding Mangrove or
+  - the updated offer needs less provision, than it already has. Meaning the user gets credited the difference.
+- When retracting an offer and deprovisioning it.Meaning the user gets credited the provision that was locked by the offer.
+- When an offer fails. The remaining provision gets credited back to the maker
+
+A challenge for an indexer is to know how much provision each offer locks. With the current events, an indexer is going to have to know the liveness, gasreq and gasprice of the offer. If the offer is not live, then also know if it has been deprovisioned. And know what the gasbase of the offerlist was when the offer was posted. With this information an indexer can calculate the exact provision locked by the offer.
+
+Another information that an indexer cannot deduce, is in what scenario the credit event happened. E.g. We don't know if the credit event happened because an offer failed or if the user simply funded Mangrove.
```

Comment:
```suggestion
The indexer also cannot deduce what scenario the credit event happened. E.g., we don't know if the credit event happened because an offer failed or because the user simply funded Mangrove.
```